File: app/src/main/java/org/blitzortung/android/dialogs/QuickSettingsDialog.java
package org.blitzortung.android.dialogs;

import android.app.Dialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Spinner;

import org.blitzortung.android.app.Main;
import org.blitzortung.android.app.R;
import org.blitzortung.android.app.view.PreferenceKey;

public class QuickSettingsDialog extends DialogFragment {

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        final LayoutInflater layoutInflater = getActivity().getLayoutInflater();

        final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());

        final String[] regionValues = getResources().getStringArray(R.array.regions_values);
        final String currentRegionValue = preferences.getString(PreferenceKey.REGION.toString(), regionValues[0]);
        int selectedRegion = getSelectedIndex(regionValues, currentRegionValue);

        final String[] rasterSizeValues = getResources().getStringArray(R.array.raster_size_values);
        final String currentRasterSizeValue = preferences.getString(PreferenceKey.RASTER_SIZE.toString(), rasterSizeValues[0]);
        int selectedRasterSize = getSelectedIndex(rasterSizeValues, currentRasterSizeValue);

        final String[] queryPeriodValues = getResources().getStringArray(R.array.query_period_values);
        final String currentQueryPeriodValue = preferences.getString(PreferenceKey.QUERY_PERIOD.toString(), queryPeriodValues[0]);
        int selectedQueryPeriod = getSelectedIndex(queryPeriodValues, currentQueryPeriodValue);

        final View view = layoutInflater.inflate(R.layout.quick_settings_dialog, null);
        final ListView selectedAreaList = (ListView) view.findViewById(R.id.selected_area);
        final String[] regions = getResources().getStringArray(R.array.regions);
        ArrayAdapter<String> ad=new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_single_choice, regions);

        selectedAreaList.setAdapter(ad);
        selectedAreaList.setChoiceMode(ListView.CHOICE_MODE_SINGLE);
        selectedAreaList.setItemChecked(selectedRegion, true);

        final Spinner rasterSizeSpinner = (Spinner) view.findViewById(R.id.selected_raster_size);
        rasterSizeSpinner.setSelection(selectedRasterSize);

        final Spinner queryPeriodSpinner = (Spinner) view.findViewById(R.id.selected_query_period);
        queryPeriodSpinner.setSelection(selectedQueryPeriod);

        builder.setView(view)
                .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        final String regionValue = regionValues[selectedAreaList.getCheckedItemPosition()];
                        final String rasterSizeValue = rasterSizeValues[rasterSizeSpinner.getSelectedItemPosition()];
                        final String queryPeriodValue = queryPeriodValues[queryPeriodSpinner.getSelectedItemPosition()];

                        Log.i(Main.LOG_TAG, "region: " + regionValue);
                        preferences.edit()
                                .putString(PreferenceKey.REGION.toString(), regionValue)
                                .putString(PreferenceKey.RASTER_SIZE.toString(), rasterSizeValue)
                                .putString(PreferenceKey.QUERY_PERIOD.toString(), queryPeriodValue)
                                .apply();
                    }
                })
                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                });

        return builder.create();
    }

    private int getSelectedIndex(String[] regionValues, String currentRegionValue) {
        int selectedRegion = 0;
        for (String regionValue : regionValues) {
            if (regionValue.equals(currentRegionValue)) {
                break;
            }
            selectedRegion++;
        }
        return selectedRegion < regionValues.length ? selectedRegion : 0;
    }
}

File: app/src/main/java/org/blitzortung/android/dialogs/QuickSettingsDialog.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.dialogs

import android.annotation.SuppressLint
import android.app.AlertDialog.Builder
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.SharedPreferences
import android.content.res.Resources
import android.os.Bundle
import android.view.View
import android.widget.Spinner
import androidx.core.content.edit
import androidx.fragment.app.DialogFragment
import androidx.preference.PreferenceManager
import org.blitzortung.android.app.R
import org.blitzortung.android.app.view.PreferenceKey
import org.blitzortung.android.app.view.get
import org.blitzortung.android.settings.putString

class QuickSettingsDialog : DialogFragment() {
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = Builder(activity)
        val layoutInflater = requireActivity().layoutInflater

        @SuppressLint("InflateParams")
        val view = layoutInflater.inflate(R.layout.quick_settings_dialog, null, false)

        val preferences = PreferenceManager.getDefaultSharedPreferences(activity as Context)

        val spinnerManager = SpinnerManager(resources, view, preferences)

        spinnerManager.add(
            R.array.regions_values,
            PreferenceKey.REGION,
            R.id.selected_region,
            defaultIndex = 1,
        )

        spinnerManager.add(
            R.array.grid_size_values,
            PreferenceKey.GRID_SIZE,
            R.id.selected_grid_size,
        )

        spinnerManager.add(
            R.array.count_threshold_values,
            PreferenceKey.COUNT_THRESHOLD,
            R.id.selected_count_threshold,
        )

        spinnerManager.add(
            R.array.query_period_values,
            PreferenceKey.QUERY_PERIOD,
            R.id.selected_query_period,
            defaultIndex = 2,
        )

        spinnerManager.add(
            R.array.interval_duration_values,
            PreferenceKey.INTERVAL_DURATION,
            R.id.selected_interval_duration,
            defaultIndex = 1,
        )

        spinnerManager.add(
            R.array.historic_timestep_values,
            PreferenceKey.HISTORIC_TIMESTEP,
            R.id.selected_historic_timestep,
            defaultIndex = 1,
        )

        spinnerManager.add(
            R.array.animation_interval_duration_values,
            PreferenceKey.ANIMATION_INTERVAL_DURATION,
            R.id.selected_animation_interval_durations,
            defaultIndex = 1,
        )

        return builder.setView(view)
            .setPositiveButton(R.string.ok) { _: DialogInterface, _: Int -> spinnerManager.updateSettings() }
            .setNegativeButton(R.string.cancel) { _: DialogInterface, _: Int -> }
            .create()
    }
}

class SpinnerManager(
    val resources: Resources,
    val view: View,
    val preferences: SharedPreferences,
) {
    private val entries = mutableListOf<Pair<PreferenceKey, () -> String?>>()

    fun add(
        valuesId: Int,
        preferenceKey: PreferenceKey,
        viewId: Int,
        defaultIndex: Int = 0,
    ) {
        val values = resources.getStringArray(valuesId)
        val currentValue = preferences.get(preferenceKey, values[defaultIndex])
        val selectedIndex = getSelectedIndex(values, currentValue)
        val spinner: Spinner = view.findViewById(viewId)
        spinner.setSelection(selectedIndex)
        entries.add(Pair(preferenceKey) { values[spinner.selectedItemPosition] })
    }

    private fun getSelectedIndex(
        values: Array<String>,
        currentValue: String,
    ): Int {
        var selectedIndex = 0
        for (regionValue in values) {
            if (regionValue == currentValue) {
                break
            }
            selectedIndex++
        }
        return if (selectedIndex < values.size) selectedIndex else 0
    }

    fun updateSettings() {
        preferences.edit {
            for (entry in entries) {
                putString(entry.first, entry.second())
            }
        }
    }
}
